runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -3585,105 +3407,46 @@ TR::Register *J9::Power::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, T
-   TR::LabelSymbol *wbLabel, *doneLabel, *startLabel, *storeLabel, *wrtBarEndLabel;
+   TR::Register *condReg, *temp1Reg, *temp2Reg, *temp3Reg, *temp4Reg;
+   TR::LabelSymbol *wbLabel, *startLabel;
    TR::RegisterDependencyConditions *conditions;
-   J9::Power::PrivateLinkage *linkage = (J9::Power::PrivateLinkage *) cg->getLinkage();
-   const TR::PPCLinkageProperties &properties = linkage->getProperties();
-   TR::Instruction *gcPoint;
-   bool stopUsingSrc = false;
 
    wbLabel = generateLabelSymbol(cg);
-   doneLabel = generateLabelSymbol(cg);
-   storeLabel = (doWrtBar) ? (generateLabelSymbol(cg)) : NULL;
 
-   tempMR1 = TR::MemoryReference::createWithRootLoadOrStore(cg, firstChild, TR::Compiler->om.sizeofReferenceAddress());
    dstReg = cg->evaluate(destinationChild);
-   TR::Register *compressedReg;
-   if (sourceChild->getReferenceCount() > 1 && (srcReg = sourceChild->getRegister()) != NULL)
-      {
-      TR::Register *tempReg = cg->allocateCollectedReferenceRegister();
-
-      // Source must be an object.
-      TR_ASSERT(!srcReg->containsInternalPointer(), "Stored value is an internal pointer");
+   srcReg = cg->evaluate(sourceChild);
 
-      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, tempReg, srcReg);
-      srcReg = tempReg;
-      stopUsingSrc = true;
-      compressedReg = srcReg;
-      if (usingCompressedPointers)
-         compressedReg = cg->evaluate(firstChild->getSecondChild());
-      }
-   else
-      {
-      srcReg = cg->evaluate(sourceChild);
-      compressedReg = srcReg;
-      if (usingCompressedPointers)
-         compressedReg = cg->evaluate(firstChild->getSecondChild());
-      }
-
-   int32_t numDeps = 11;
-   if (usingCompressedPointers)
-      numDeps++;
-   if ((gcMode == gc_modron_wrtbar_satb) || (comp->getOptions()->realTimeGC()))
-      numDeps++;
-   if (!firstChild->skipWrtBar())
-      numDeps += 2;
-
-   conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(numDeps, numDeps, cg->trMemory());
+   conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(7, 7, cg->trMemory());
    temp1Reg = cg->allocateRegister();
    temp2Reg = cg->allocateRegister();
    temp3Reg = cg->allocateRegister();
    temp4Reg = cg->allocateRegister();
    condReg = cg->allocateRegister(TR_CCR);
 
-   // !!! Adding any dependency before the baseReg, you have to be careful with the excludeGPR0 order
-   TR::addDependency(conditions, dstReg, TR::RealRegister::gr3, TR_GPR, cg);
-
-   if (comp->getOptions()->realTimeGC())
-      TR::addDependency(conditions, srcReg, TR::RealRegister::gr5, TR_GPR, cg);
-   else
-      TR::addDependency(conditions, srcReg, TR::RealRegister::gr4, TR_GPR, cg);
-
-   TR::addDependency(conditions, temp1Reg, TR::RealRegister::gr11, TR_GPR, cg);
+   TR::addDependency(conditions, dstReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   conditions->getPreConditions()->getRegisterDependency(conditions->getAddCursorForPre() - 1)->setExcludeGPR0();
```

Comment:
I think this will cause some register shuffling since before dependencies were shared between the array check store and write barrier helpers.

Reply:
I don't think removing these explicit dependencies should cause any additional register shuffles. The code generated for the write barrier comes after the label with these dependencies, meaning that the more specific dependencies from the write barrier would be encountered by the local register allocator before these ones.